### DISCUSSION TITLE
Question about running the find_optimal_representation method in the PEC module

### DISCUSSION THREAD

**Discussion by eshaspark** [unitaryfund/mitiq#1995] [Q&A] [answered]
I have a question about using the PEC module of mitiq:
> In what form does one input a basis of noisy operations for the Quasiprobability decomposition?

I'm looking to run the `find_optimal_representation` method which decomposes ideal operations to a basis of noisy operations. The basis I'd like consists of a bunch of unitaries and post selection operations. The documentation mentions the function: 
```
mitiq.pec.representations.optimal.find_optimal_representation(
    ideal_operation,
    noisy_operations,
    tol=1e-08,
    initial_guess=None,
    is_qubit_dependent=True,
)
```

Do the `noisy_operations` have to be initialized as quantum circuits? (I'm using qiskit frontend)

I tried that and it doesn't seem to be working. Also, I'd like to use post-selection operations so numerical matrices are preferred, for the noisy operation parameter. I can see there's an option to input a superoperator but it isn't clear to me whether that's sufficient.

**Comment by andreamari** [comment]
Hi @eshaspark, 
to use the function `find_optimal_representation` you need to define, as a basis, a list of `NoisyOperation` objects. Each `NoisyOperation` object is defined by the underlying circuit (e.g. a Qiskit or Cirq circuit) and the associated `channel_matrix` (the superoprator as a numerical matrix).

You can find an example on how define noisy operations at this [link](https://mitiq.readthedocs.io/en/stable/guide/pec-3-options.html#building-operationrepresentation-objects).
You can find an example no how to use `find_optimal_representation` at this [link](https://mitiq.readthedocs.io/en/stable/guide/pec-3-options.html#finding-an-optimal-operationrepresentation).

In mitiq we assume that noisy operations are noisy versions of ideal **unitary** gates, so post-selection operations are not supported in principle. However, you may try to define noisy operations with post-selection gates, an see if you get an error in the process.

Alternatively, if you have the numerical super-operator matrices and you are just interested in solving the underlying optimization problem from a numerical perspective (i.e. no circuits, no NoisyOperations objects, etc..), you may have a look at the function `minimize_one_norm` .  More info at this [link](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.pec.representations.optimal.minimize_one_norm).

**Reply by eshaspark** [reply]
Hi @andreamari,

Thanks for the explainer and the links. When defining the NoisyOperation object, I see one needs the underlying (Qiskit or cirq) circuit _as well as_ numerical superoperator matrices. Defining post selections as superoperator matrices is straightforward ✅ .
It seems to implement PEC in mitiq (not just for solving the optimization problem from a numerical perspective, but to also run PEC), this superoperator matrix isn't enough and one needs the circuits for the noisy operations as well.

I want to understand how to include underlying basis circuits for post-selections. In Qiskit, the measurement + reset gate wouldn't achieve the same as it would reset all measurements to 0, regardless of the outcome. There aren't native post-selection gates in Qiskit as it technically isn't a gate, but rather the fact that certain measurement outcomes are discarded. Is there a way to include an operation like this to run PEC with mitiq?

**Reply by andreamari** [reply, marked as answer]
You are right. If you aim to implement PEC, the NoisyOperation objects must be applicable in a circuit as actual real gates on real qubits.
So, if the specific operations you are interested in are not implementable in Qiskit, I think there is no way to apply PEC with or without Mitiq.

About Qiskit conditional operations: you might be interested in [this method](https://qiskit.org/documentation/locale/bn_BN/stubs/qiskit.circuit.InstructionSet.c_if.html). It allows to apply gates based on a previous measurement outcome.  Likely this kind of operations won't be compatible with Mitiq due to conversion errors, but at least you may find a way to manually implement your noisy operations in a Qiskit circuit.

**Reply by eshaspark** [reply]
Ok, thanks for confirming that post-selections are currently not implementable on mitiq! 

There is a second issue I encounter with `find_optimal_representation` method - when inputing circuits to the method. This issue is related to what you mean by 'real gates'. There are operations that are implementable on qiskit as gates in software, but are mitiq is unable to decompose these circuits when they are input to PEC. Pl refer to the screenshot below.

The only qiskit circuit inputs that are being accepted by mitiq without error involve native gates: h, cx, and so on. When I define an arbitary unitary and then create a [Quantum Gate](https://qiskit.org/documentation/stubs/qiskit.extensions.UnitaryGate.html) object from this unitary on qiskit, the input is not being accepted by mitiq. So it seems the PEC simulation can't be performed for arbitrary unitary circuits on qiskit (without circuit synthesis)? The numerical minimize_one_norm method works fine, but I'd like to be able to run a simulation after computing the coefficients.

![Screenshot 2023-09-14 at 12 13 49 PM](https://github.com/unitaryfund/mitiq/assets/8880428/c16a2674-3ab2-4211-9e1e-d2574c0f3f5a)

**Comment by andreamari** [comment]
You are right. Custom gates such as those defined by numerical matrices are not supported in Mitiq. So one can apply PEC only if the  circuit is already decomposed into native  gates or at least standard gates such as H, S, T, CZ, etc... .
